### PR TITLE
[libc][math] Implement double precision cbrt correctly rounded to all rounding modes.

### PR DESCRIPTION
**Pull request by lntue**
Division-less Newton iterations algorithm for cube roots.

1. **Range reduction**

For `x = (-1)^s * 2^e * (1.m)`, we get 2 reduced arguments `x_r` and `a` as:
```
  x_r = 1.m
  a   = (-1)^s * 2^(e % 3) * (1.m)
```
Then `cbrt(x) = x^(1/3)` can be computed as:
```
  x^(1/3) = 2^(e / 3) * a^(1/3).
```

In order to avoid division, we compute `a^(-2/3)` using Newton method and then
multiply the results by a:
```
  a^(1/3) = a * a^(-2/3).
```

2. **First approximation to a^(-2/3)**

First, we use a degree-7 minimax polynomial generated by Sollya to
approximate `x_r^(-2/3)` for `1 <= x_r < 2`.
```
  p = P(x_r) ~ x_r^(-2/3),
```
with relative errors bounded by:
```
  | p / x_r^(-2/3) - 1 | < 1.16 * 2^-21.
```

Then we multiply with `2^(e % 3)` from a small lookup table to get:
```
  x_0 = 2^(-2*(e % 3)/3) * p
      ~ 2^(-2*(e % 3)/3) * x_r^(-2/3)
      = a^(-2/3)
```
with relative errors:
```
  | x_0 / a^(-2/3) - 1 | < 1.16 * 2^-21.
```
This step is done in double precision.

3. **First Newton iteration**

We follow the method described in:
  Sibidanov, A. and Zimmermann, P., "Correctly rounded cubic root evaluation
  in double precision", https://core-math.gitlabpages.inria.fr/cbrt64.pdf
to derive multiplicative Newton iterations as below:
Let `x_n` be the nth approximation to `a^(-2/3)`.  Define the n^th error as:
```
  h_n = x_n^3 * a^2 - 1
```
Then:
```
  a^(-2/3) = x_n / (1 + h_n)^(1/3)
           = x_n * (1 - (1/3) * h_n + (2/9) * h_n^2 - (14/81) * h_n^3 + ...)
```
using the Taylor series expansion of `(1 + h_n)^(-1/3)`.

Apply to `x_0` above:
```
  h_0 = x_0^3 * a^2 - 1
      = a^2 * (x_0 - a^(-2/3)) * (x_0^2 + x_0 * a^(-2/3) + a^(-4/3)),
```
it's bounded by:
```
  |h_0| < 4 * 3 * 1.16 * 2^-21 < 2^-17.
```
So in the first iteration step, we use:
```
  x_1 = x_0 * (1 - (1/3) * h_n + (2/9) * h_n^2 - (14/81) * h_n^3)
```
Its relative error is bounded by:
```
  | x_1 / a^(-2/3) - 1 | < 35/242 * |h_0|^4 < 2^-70.
```
Then we perform Ziv's rounding test and check if the answer is exact.
This step is done in double-double precision.

4. **Second Newton iteration**

If the Ziv's rounding test from the previous step fails, we define the error
term:
```
  h_1 = x_1^3 * a^2 - 1,
```
And perform another iteration:
```
  x_2 = x_1 * (1 - h_1 / 3)
```
with the relative errors exceed the precision of double-double.
We then check the Ziv's accuracy test with relative errors < 2^-102 to
compensate for rounding errors.

5. **Final iteration**
 
If the Ziv's accuracy test from the previous step fails, we perform another
iteration in 128-bit precision and check for exact outputs.

### CODE REVIEW

**Comment by lntue** [comment]
@zimmermann6

**Comment by zimmermann6** [comment]
all tests are ok on my side

File: libc/src/math/generic/CMakeLists.txt [hunk 1]
```diff
@@ -4180,3 +4180,19 @@ add_entrypoint_object(
...
+  HDRS
+    ../cbrt.h
+  COMPILE_OPTIONS
+    -O3
+  DEPENDS
+    libc.hdr.fenv_macros
+    libc.src.__support.FPUtil.fenv_impl
+    libc.src.__support.FPUtil.fp_bits
+    libc.src.__support.FPUtil.multiply_add
+    libc.src.__support.macros.optimization
```
Comment (overmighty) [inline libc/src/math/generic/CMakeLists.txt:4197]: Should also depend on `polyeval`, `double_double`, `dyadic_float`, `integer_literals`.

File: libc/src/math/generic/cbrt.cpp [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+#include "src/__support/macros/optimization.h" // LIBC_UNLIKELY
+
+#if ((LIBC_MATH & LIBC_MATH_SKIP_ACCURATE_PASS) != 0)
+#define LIBC_MATH_CBRT_SKIP_ACCURATE_PASS
+#endif
+
+namespace LIBC_NAMESPACE_DECL {
+
+using DoubleDouble = fputil::DoubleDouble;
+using Float128 = typename fputil::DyadicFloat<128>;
```
Comment (overmighty) [inline libc/src/math/generic/cbrt.cpp:29]: `typename` isn't necessary here, is it?

File: libc/src/math/generic/cbrt.cpp [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+// Polynomial generated by Sollya with:
+// > P = fpminimax(x^(-2/3), 7, [|D...|], [1, 2]);
+// > dirtyinfnorm(P/x^(-2/3) - 1, [1, 2]);
+// 0x1.28...p-21
+constexpr double intial_approximation(double x) {
+  constexpr double COEFFS[8] = {
+      0x1.bc52aedead5c6p1,  -0x1.b52bfebf110b3p2,  0x1.1d8d71d53d126p3,
+      -0x1.de2db9e81cf87p2, 0x1.0154ca06153bdp2,   -0x1.5973c66ee6da7p0,
+      0x1.07bf6ac832552p-2, -0x1.5e53d9ce41cb8p-6,
+  };
```
Comment (overmighty) [inline libc/src/math/generic/cbrt.cpp:43]: Do we have a policy on when to use C-style arrays, and when to use `cpp::array`?

File: libc/src/math/generic/cbrt.cpp [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+// === Step 5 - Final iteration ===
+// If the Ziv's accuracy test from the previous step fails, we perform another
+// iteration in 128-bit precision and check for exact outputs.
+//
+// TODO: It is possible to replace this costly computation step with special
+// exceptional handling, similar to what was done in the CORE-MATH project:
+// https://gitlab.inria.fr/core-math/core-math/-/blob/master/src/binary64/cbrt/cbrt.c
+
+LLVM_LIBC_FUNCTION(double, cbrt, (double x)) {
+  using FPBits = typename fputil::FPBits<double>;
```
Comment (overmighty) [inline libc/src/math/generic/cbrt.cpp:146]: I don't think `typename` is necessary here either.

File: libc/src/math/generic/cbrt.cpp [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+
+  uint64_t x_u = FPBits(x).uintval();
+  uint64_t x_abs = x_u & 0x7fff'ffff'ffff'ffff;
```
Comment (overmighty) [inline libc/src/math/generic/cbrt.cpp:149]: There's several instances of masking with hex literals in this file. It's done in other higher math function implementations too, so maybe it's fine, I'm not sure.

File: libc/src/math/generic/cbrt.cpp (hunk 3 above)
Comment (lntue) [reply libc/src/math/generic/cbrt.cpp:43]: I don't think we have one.

File: libc/src/math/generic/CMakeLists.txt (hunk 1 above)
Comment (lntue) [reply libc/src/math/generic/CMakeLists.txt]: Done.

File: libc/src/math/generic/cbrt.cpp (hunk 2 above)
Comment (lntue) [reply libc/src/math/generic/cbrt.cpp]: Done.

**Review by overmighty — APPROVED** [review]

**Review by SchrodingerZhu — APPROVED** [review]

**Pull request merged**